### ISSUE TITLE
Connection between Cryo-stabilised Flux Duct [Thermal Dynamics] and aluminium wire

### ISSUE THREAD

**Issue by Brycey92** [micdoodle8/Galacticraft#1544]
Just today, my server started crashing repeatedly every few minutes with this crash report: https://gist.github.com/Brycey92/4f8b71cddde7b00ed727

I have no idea what block causes this problem, nor how to reproduce the issue. I am running Galacticraft Core and Planets 3.0.11.310.

**Comment by radfast** [comment]
Hmm, that's not part of Galacticraft which has changed for maybe 2-3 months.  I can see it is our code crashing, but right now I don't understand __why__ (and I would like to, as this could be relevant to other situations).  I can fix this without understanding why, so don't worry we will get this fixed, but I'd really like to try to figure this out.  A few questions that could help me please:

* Which version of Galacticraft were you running before today?

* Did you install / update any other new mods recently?

* Did you change your Java settings recently when running Minecraft (I noticed ```-XX:+UseFastAccessorMethods -XX:ParallelGCThreads=4 ``` which are unusual, though I guess performance related - please can you test it without those?)

*  Do you have any mods installed which try to make Minecraft multithreaded - that's an impressive modpack you have but I'm not familiar with all of them.  (I appreciate that if you're not a coder you may not be able to answer this, but a quick text search of your .minecraft/configs folder for the word "thread" might find something?)

The block causing the problem is a connection between a GC Aluminium Wire and probably a machine from another mod which the wire is powering.   You have a list of the players online in that crash report, and their locations, so it must be something in or near one of their bases:

'KilledAssassin'/5, l='world', x=-2930.70, y=64.00, z=-4223.83
'brainbendjr'/53, l='world', x=-778.03, y=95.99, z=-429.88
'StormSire'/54, l='world', x=-2937.60, y=64.00, z=-4218.43
'Shadowbrine'/104, l='world', x=-843.30, y=75.03, z=-504.32
'reconTHEgoddess'/446, l='world', x=-910.76, y=79.76, z=-525.47
'jjkidd'/557, l='world', x=1143.34, y=91.00, z=4082.95

I'll also keep looking at this from our end, see what I can find.

**Comment by Brycey92** [comment]
I was running 3.0.10.298 before today, but these crashes happened in this version as well ,which was what prompted my update.

No other mods were changed in about a week before this issue started. Many players have started using aluminum wires to convert from RF/Mekanism energy to IC2, since MFR RedNet Energy Cables have begun misbehaving.

If removing aluminum wires does not resolve the issue, I will test it without the performance arguments.

We have many mods and many plugins, and a lot have multithreading, though none attempt make the base game multithreaded.

I will investigate the bases of the players you listed, remove the aluminum wires from them, and report back.

**Comment by radfast** [comment]
Please try running this special build on your server: http://ci.micdoodle8.com/job/Galacticraft-1.7/311/
(the players can still run 3.0.11.310, that is absolutely 100% compatible).

What this special build __should__ do (I hope) is not to crash at that wire / machine connection, but instead spam your log with a "Galacticraft SEVERE" error message which gives you the exact coordinates of the problem block.  (Actually it gives coordinates for two blocks, one before and one after the problem, as I'm not sure which one it is right now.)  Hopefully this will work like that but I can't be 100% sure because I don't have your world + pack to test it in.

If this works out as I hope, what I'll ask you to do is tp to the coordinates which it marks as the problem, and see what block is there - it will most likely be a machine from another mod.  (Red faces if it is a machine from Galacticraft)

Thanks for helping.

**Comment by radfast** [comment]
My bad: please use this special build, not 311:  http://ci.micdoodle8.com/job/Galacticraft-1.7/312/

**Comment by Brycey92** [comment]
Removing the aluminum wires from players' bases stopped the crashes, but adding them back with build 312 server-side produced no crashes or warnings in the console.

**Comment by radfast** [comment]
The debug information will be in the server log, not the console.  (We don't want to spam your console :) )

I would very much appreciate it if you can follow this through, because this indicates there is probably some more fundamental compatibility problem.  I still think the most likely cause is the aluminium wire connecting to a specific machine, almost certainly a machine from another mod (and something fairly rare or recent, otherwise another server would have reported this issue long ago).

**Comment by Brycey92** [comment]
I discovered my console full of errors today: http://i.imgur.com/JqBUmps.png

Is this the aluminum wire crash prevention kicking in or is it another mod? Either way, I'll check out the coordinates later today.

**Comment by radfast** [comment]
That's the alu wire crash prevention.  Please let us know what is at those coordinates - name of block and a screenshot would be great, and then I think we will have an answer and can remove the debug message.

**Comment by radfast** [comment]
It's odd that the same coordinates are in the list twice, maybe the player made an energy loop with the same wiring network making two connections to the block?  I feel like a couple of screenshots will probably resolve this.

**Comment by Brycey92** [comment]
Here is a screenshot of the area causing the issue: http://i.imgur.com/nhdTsJX.png
The block referred to by the coordinates is the cryo-stabilized fluxduct directly connected to the aluminum wire and ultimate universal cable.

**Comment by radfast** [comment]
Wow.  That's an evil setup.  Players are not supposed to connect wires to other mods' wires, for a start - I thought we had blocked most of them, but obviously not the "cryo-stabilized flux duct"...

OK thank you this is excellent, that gives me the info needed to test and try to reproduce the issue on my test platforms.

**Comment by Brycey92** [comment]
Without the aluminum wire, my players will be unable to (reliably) convert power between IC2 and other mods. Is it possible to fix the problem at its source rather than blacklist fluxducts?

**Comment by radfast** [comment]
I am still looking at the issue, but for sure the first thing I will do is try to identify the exact cause of the error.

Fluxducts will certainly be blacklisted for direct "wire to wire" connections.  The reason those are blacklisted is because if you run 2 wires side by side, it creates loops - maybe loops at multiple points - which can cause lag if the other mod is badly coded.  But energy conversion to / from IC2 using Galacticraft is still permitted: just use a Galacticraft block, for example an Energy Storage Module.

Alternatvely, why don't your players use Galacticraft wiring throughout their bases instead of the fluxducts?  Galacticraft wiring might not look as cool as that, but it works reliably (apart from this specific issue), it connects to everything and it is also high performance - meaning a complex wiring network should make for less server lag if Galacticraft wires are used than wires from any other mod.  Most server owners do not seem to understand that in large bases, a big proportion of the lag comes from the wiring and pipes, and not from the actual generators or power-consuming machines.

**Comment by Brycey92** [comment]
They protest that aluminum wires do not have the infinite throughput that cryo-stabilized fluxducts have. Several of my players have also been exploiting power dupe bugs with them. Although I have been unable to determine the exact source of these dupes, I can build and screenshot a setup they had to produce the dupe bug.

**Comment by radfast** [comment]
Yes please, re the power dupe bug.  If it involves interaction with other mods it may not be Galacticraft causing it - I'll keep an open mind until I see it - but even so we may be able to do something to detect and stop it.  Please also post your ```power-gc3.conf``` settings.

I don't understand about infinite throughput.  Galacticraft __wires__ have infinite throughput.  (Galacticraft __machines__ (including Energy Storage Module / Cluster) do not have infinite throughput.)  But anything the cryo-stabilized fluxducts can do, the aluminium wires should be able to do, and probably better for your server and players: less lag, and more accurate resumption of operation after chunk loading.

Put it this way: if the aluminium wires are __not__ the best available wires then I'll be looking at what is necessary to make them the best.

**Comment by Brycey92** [comment]
As evidenced by some tests one of my admins and I ran a few weeks ago, aluminum wires do not provide infinite throughput, though they do provide a lot. We measured it to be less throughput than Mekanism ultimate universal cables, which aren't infinite themselves either.

Here are the two versions of my power-gc3.conf file. The first section is the default config values, and the second section is our current revised config to match the conversion ratios employed by Mekanism in an attempt to stop the aluminum wire power dupe. I have not yet tested if my modifications resolve the issue. https://gist.github.com/Brycey92/f4f0b08bd78117b8b19a

On another note, have you spoken to the Mekanism, Thermal Dynamics, IC2, and MFR devs about implementing your performance tweaks in their mods?

**Comment by radfast** [comment]
Ok, there can always be power duplication (meaning a power loop where power increases as it goes around) if the player uses three mod transititions: e.g. GC->Mek, Mek->IC2 and IC2->GC, __if__ your configs are not set up in a matching way.  That's kind of obvious.  wiki.micdoodle8.com/wiki/Energy#Conversion_Ratios

If this is the reason for the 'dupe' then its not a bug in the aluminium wire, it's just you needed better configs.  (I would assume it is one of the first things which any high quality modpack aims to get right.)

We also include in Galacticraft an optional power conversion loss factor which you can use, a loss of 10% to 20% should normally be enough to stop any dupe which you cannot stop in another way (unless your conversion factors in other mods are way off).  http://wiki.micdoodle8.com/wiki/Energy#Optional_lossy_conversion

**Comment by radfast** [comment]
> As evidenced by some tests one of my admins and I ran a few weeks ago, aluminum wires do not provide infinite throughput, though they do provide a lot. We measured it to be less throughput than Mekanism ultimate universal cables, which aren't infinite themselves either.

There is no hard-coded throughput limit in the alu wires, and no reason why it should be limited for a network of alu wires.  The only limit is that when alu wire is putting power into __IC2 machines__ power transfer _per machine_ is capped at 128EU/t for standard Alu wire, and 256EU/t for heavy - that was a design decision so as not to make the highest tier IC2 wiring obsolete, because IC2 modpacks tend to want to use IC2 wiring.  Could be that is what you were seeing?  It's hard to say without seeing your exact test setup.  That cap should not affect power transfer into Mekanism or any MFR machines or other RF-based machines, for example you could run a 10 LaserDrill setup from regular aluminium wiring with no problems.

(We actually plan to code in some 'problems' in future in Galacticraft hard mode, as right now the aluminium wire is a little OP.  http://wiki.micdoodle8.com/wiki/Game_Balance#Hard_Mode)

**Comment by Brycey92** [comment]
The power dupe only used TD fluxducts and GC aluminum wires, and TD does not have a power ratio config. If we are unable to determine the source of the dupe, though, I will take your advice and implement a small loss in GC wire power conversion.

As for the IC2 machine / GC wire power transfer limit, we did not anticipate such a feature, and had been primarily testing with IC2 machines. How would one raise the amount of power that could be transferred from a GC wire network to an IC2 machine? Would a simple MV or HV transformer suffice or would it have to be something more complicated?

As for the nerfing of alu wires, would it be possible to nerf the existing ones but implement progressively more expensive and more versatile ones, like other mods do?

You also never mentioned if you spoke to other devs about implementing your performance boosts in their own mods.

**Comment by dj3520** [comment]
> How would one raise the amount of power that could be transferred from a GC wire network to an IC2 machine?

This is hard coded. What you could do is convert it via Energy Storage Module or Cluster then use IC2 wires to transfer the energy from the module/cluster to the machines. If the output coming out with the cluster isn't enough, you'll need to hook up more in parallel.

>Would a simple MV or HV transformer suffice

This simply changes IC2 energy mechanics. Using them with GC wires does nothing. For GC, power is power.

>implement progressively more expensive and more versatile ones

The original reason for nerfing them is so that IC2 wires stayed more valuable. The limit from what I understand is per machine, not wire. If you need more throughput for EU and EU only, use IC2 wires.

@radfast I use GC wires to transfer energy from my Big Reactor turbines (RF) on my server to IC2 MFSUs (EU) and I don't see any limits. It's not a problem since otherwise there wouldn't be nearly enough throughput (~40k gJ/t?) but I figure you should know that. Perhaps a harder recipe should be a config option for modpack makers, or trigger if another mod with wires exists. I think they're pretty valuable since they convert different energies, especially since if they didn't exist, the only other way would be ESMs or ESCs.

**Comment by radfast** [comment]
Thank you @dj3520 you have covered all the points I was going to make.

@Brycey92  please provide further information of this:
> The power dupe only used TD fluxducts and GC aluminum wires

What blocks do I connect to see the dupe?

**Comment by Brycey92** [comment]
Actually @dj3520, the reason for nerfing the wires, according to the link, is to encourage the use of GC beams.

Regardless, I support your request for configurable harder recipes, since the current ones are extremely cheap

**Comment by Brycey92** [comment]
@radfast the setup was a checker-board mesh-type cube alternating fluxducts and alu wires with something inside, but I neither remember the contents of the cube nor which type of cable was connected to which sides of the blocks in the center. I will reproduce and explain the setup as soon as I am able to ask the player who created it.

**Comment by radfast** [comment]
Right, I've fixed the original issue (fix will be in build .321).  It was caused by the wire-to-wire connection between the alu wire and the flux duct, and the following rare set of circumstances which are very specific to this player's build:

* the flux duct transmitted energy in tick N, and then it ran out of energy in tick N+1 (nothing left at the power source supplying the flux duct)

* the alu wire did not put all of its energy into the IC2 machine at tick N, because the IC2 machine was given some or all of the energy it needed by the Mekanism Universal Cable also connected to it

* on tick N+1, the alu wire put some of the left-over energy back into the connected flux duct (because the flux duct was by now not transmitting anything to the alu wire, the flux duct became something which can accept energy)

* when the energy was put into the flux duct from the alu wire, the flux duct _at that point started transmitting energy_ into the alu wire, which is what messed things up

I've fixed this by a small change in the alu wire.

Your reporting this issue was helpful, because the same problem could in theory be caused by any machine from any mod which has the capacity to input and output energy on the same side (for example, a Tesseract) - though it must be an __rare__ problem because it depends on the very specific setup conditions explained above, and that's why nobody has reported it before.  (Galacticraft's wiring code has been stable for around 9 months now.)

**Comment by radfast** [comment]
I cannot investigate the suggesting duping issue without an exact description.

Note that some of the RC versions of Thermal Dynamics had a power duping issue in Cryo-Stabilised Flux Ducts, see https://github.com/CoFH/Feedback/issues/264

**Comment by Brycey92** [comment]
Thanks for the fix! I can finally get my console back to normal.

As for TD, we're using a stable version, and the dupe was done with some other fluxduct, not the cryo-stabilized one, but thanks for looking into it for me. I suppose the best way to go about the power dupe problem would be to open a new issue and reference this one. I'll do that as soon as I figure out how to reproduce it.

**Comment by radfast** [comment]
@Brycey92  you can post the duping thing on this issue, I (and all subscribers) get notified of your posts here the same way as if you started a new issue.  It might also help keep it on the DL, if the heading of the issue is __not__ about duping.  There's a type of player who actually searches issues lists for duping reports when figuring out what might be possible in a modpack.
